File: app/SkillLoader.ts
import fs from "fs";
import path from "path";

export class Skill {
  constructor(
    public name: string,
    public description: string,
  ) {}
}

export class SkillLoader {
  SkillDir: string = "skills";
  skills: Skill[] = [];
  mapContent: Map<string, Skill & { content: string }> = new Map();

  constructor() {
    this.getSkills();
  }

  /** 同步加载 skills/ 目录下所有 SKILL.md */
  private getSkills() {
    const skillsDir = path.resolve(process.cwd(), this.SkillDir);

    if (!fs.existsSync(skillsDir)) {
      console.warn(`[SkillLoader] skills 目录不存在: ${skillsDir}`);
      return;
    }

    // 递归扫描所有 SKILL.md 文件
    const entries = this.walkDir(skillsDir).filter(
      (f) => path.basename(f) === "SKILL.md",
    );

    for (const entry of entries) {
      const content = fs.readFileSync(entry, "utf-8");

      // 解析 YAML front matter：name 和 description
      const nameMatch = content.match(/^name:\s*(.*)$/m);
      const descMatch = content.match(/^description:\s*(.*)$/m);

      const name = nameMatch?.[1]?.trim() ?? null;
      const description = descMatch?.[1]?.trim() ?? null;

      if (name && description) {
        this.skills.push(new Skill(name, description));
        this.mapContent.set(name, { name, description, content });
      }
    }

    console.log(`[SkillLoader] 加载了 ${this.skills.length} 个技能`);
  }

  /** 递归遍历目录，返回所有文件路径 */
  private walkDir(dir: string): string[] {
    const result: string[] = [];
    const list = fs.readdirSync(dir, { withFileTypes: true });
    for (const item of list) {
      const fullPath = path.join(dir, item.name);
      if (item.isDirectory()) {
        result.push(...this.walkDir(fullPath));
      } else {
        result.push(fullPath);
      }
    }
    return result;
  }

  getSkillByName(name: string) {
    return this.mapContent.get(name);
  }
}

File: skill_loader.py
import os
import re
from dataclasses import dataclass, field


@dataclass
class Skill:
    name: str
    description: str


@dataclass
class SkillWithContent(Skill):
    content: str = ""


class SkillLoader:
    """从 skills/ 目录加载所有 SKILL.md 文件"""

    SkillDir: str = "skills"
    skills: list[Skill] = field(default_factory=list)
    _map_content: dict[str, SkillWithContent] = field(default_factory=dict)

    def __init__(self):
        self.skills = []
        self._map_content = {}
        self._get_skills()

    def _get_skills(self) -> None:
        """同步加载 skills/ 目录下所有 SKILL.md"""
        # skills 目录在项目根目录
        skills_dir = os.path.join(
            os.path.dirname(os.path.dirname(os.path.abspath(__file__))),
            self.SkillDir,
        )

        if not os.path.exists(skills_dir):
            print(f"[SkillLoader] skills 目录不存在: {skills_dir}")
            return

        entries = self._walk_dir(skills_dir)
        entries = [f for f in entries if os.path.basename(f) == "SKILL.md"]

        for entry in entries:
            with open(entry, "r", encoding="utf-8") as f:
                content = f.read()

            name_match = re.search(r"^name:\s*(.*)$", content, re.MULTILINE)
            desc_match = re.search(r"^description:\s*(.*)$", content, re.MULTILINE)

            name = name_match.group(1).strip() if name_match else None
            description = desc_match.group(1).strip() if desc_match else None

            if name and description:
                self.skills.append(Skill(name=name, description=description))
                self._map_content[name] = SkillWithContent(
                    name=name, description=description, content=content
                )

        print(f"[SkillLoader] 加载了 {len(self.skills)} 个技能")

    def _walk_dir(self, directory: str) -> list[str]:
        """递归遍历目录，返回所有文件路径"""
        result: list[str] = []
        for item in os.listdir(directory):
            full_path = os.path.join(directory, item)
            if os.path.isdir(full_path):
                result.extend(self._walk_dir(full_path))
            else:
                result.append(full_path)
        return result

    def get_skill_by_name(self, name: str) -> SkillWithContent | None:
        return self._map_content.get(name)
